apex/contrib/csrc/optimizers/multi_tensor_distopt_adam_kernel.cu

```diff
@@ -14,151 +14,305 @@
 #define ILP 4
 
 template<typename T>
-__device__ __forceinline__ bool is_aligned(T* p){
+__device__ __forceinline__ bool is_aligned(const T* p){
   return ((uint64_t)p) % (ILP*sizeof(T)) == 0;
 }
 
 template<typename T>
-__device__ __forceinline__ void load_store(T* dst, T* src, int dst_offset, int src_offset){
+__device__ __forceinline__ void load_store(
+  T* dst,
+  const T* src,
+  int dst_offset = 0,
+  int src_offset = 0){
   typedef typename std::aligned_storage<ILP*sizeof(T), ILP*alignof(T)>::type LT;
-  ((LT*)dst)[dst_offset] = ((LT*)src)[src_offset];
+  ((LT*)dst)[dst_offset] = ((const LT*)src)[src_offset];
 }
 
 typedef enum{
-  ADAM_MODE_0   =0, // eps under square root
-  ADAM_MODE_1   =1  // eps outside square root
+  ADAM_MODE_0   =0, // L2 regularization mode
+  ADAM_MODE_1   =1  // Decoupled weight decay mode(AdamW)
 } adamMode_t;
 
-template <int DEPTH, typename T, typename GRAD_T>
+/* Multi-tensor Adam
+ *
+ * Updates params in-place and outputs a copy with a desired datatype.
+ */
+template <typename T, typename GRAD_T, typename PARAM_OUT_T>
 struct DistAdamFunctor
 {
+  // Vectorized local compute
+  __device__ __forceinline__ static void local_step(
+    T p[ILP],
+    T m[ILP],
+    T v[ILP],
+    const GRAD_T g[ILP],
+    const T grad_scale,
+    const float beta1,
+    const float beta2,
+    const float beta1_correction,
+    const float beta2_correction,
+    const float eps,
+    const float lr,
+    adamMode_t mode,
+    const float weight_decay) {
+    if (mode == ADAM_MODE_0) { // L2
+#pragma unroll
+      for (int ii = 0; ii < ILP; ii++) {
+        T scaled_grad = (g[ii] * grad_scale) + (weight_decay * p[ii]);
+        m[ii] = beta1*m[ii] + (1-beta1)*scaled_grad;
+        v[ii] = beta2*v[ii] + (1-beta2)*scaled_grad*scaled_grad;
```

Comment:
https://developer.nvidia.com/blog/lerp-faster-cuda/ can be applied here

Reply:
That's a neat trick, I'll make the change.